src/twisted/application/internet.py

```diff
@@ -761,31 +772,63 @@ def _cancelConnectWaiters(self):
         """
         self._unawait(Failure(CancelledError()))
 
+    @_machine.output()
+    def _ignoreAndCancelConnectWaiters(self, f):
+        """
+        Notify all pending requests for a connection that no more connections
+        are expected, after ignoring the Failure passed in.
+        """
+        self._unawait(Failure(CancelledError()))
+
 
     @_machine.output()
     def _finishStopping(self):
         """
         Notify all deferreds waiting on the service stopping.
         """
+        self._doFinishStopping()
+
+    @_machine.output()
+    def _ignoreAndFinishStopping(self, f):
+        """
+        Notify all deferreds waiting on the service stopping, and ignore the
+        Failure passed in.
+        """
+        self._doFinishStopping()
+
+    def _doFinishStopping(self):
         self._stopWaiters, waiting = [], self._stopWaiters
         for w in waiting:
             w.callback(None)
 
 
     @_machine.input()
-    def whenConnected(self):
+    def whenConnected(self, failAfterFailures=None):
         """
         Retrieve the currently-connected L{Protocol}, or the next one to
         connect.
 
+        @param failAfterFailures: number of connection failures after which
+        the Deferred will deliver a Failure (None means the Deferred will
```

Comment:
Convention is to indent here, as https://github.com/glyph/python-docstring-mode would.